### PR TITLE
Client Downlink Demodulation and Decoding 

### PR DESCRIPTION
**Pull request by rdoosty**
This PR:
- Adds demodulation and decoding to client downlink
- Adds bit and block error calculation
- Fixes downlink beamforming scaling in Agora

### CODE REVIEW

File: src/agora/dozf.cpp [hunk 1]
```diff
@@ -82,6 +82,10 @@ void DoZF::compute_precoder(const arma::cx_fmat& mat_csi,
             mat_dl_zf = mat_ul_zf * arma::inv(mat_calib);
         } else
             mat_dl_zf = mat_ul_zf;
+	// We should scaling the beamforming matrix
+	// So the IFFT output can scaled with OFDM_CA_NUM
+	// See Argos paper (Mobicom 2012) Sec. 3.4
+	mat_dl_zf /= abs(mat_dl_zf).max();
```
Comment (jianding17) [inline src/agora/dozf.cpp:88]: These lines should be aligned with line 84. There are also some typos in the comment.

File: src/agora/agora.cpp [hunk 2]
```diff
@@ -1195,13 +1196,13 @@ void Agora::save_tx_data_to_file(UNUSED int frame_id)
     FILE* fp = fopen(filename.c_str(), "wb");
 
     for (size_t i = 0; i < cfg->dl_data_symbol_num_perframe; i++) {
-        int total_data_symbol_id
-            = (frame_id % TASK_BUFFER_FRAME_NUM) * cfg->data_symbol_num_perframe
-            + i + cfg->dl_data_symbol_start;
+        size_t total_data_symbol_id = (frame_id % TASK_BUFFER_FRAME_NUM)
+                * cfg->dl_data_symbol_num_perframe
+            + i;
```
Comment (jianding17) [inline src/agora/agora.cpp:1201]: Can you use `cfg->get_total_data_symbol_idx_dl(frame_id, i)` here?

File: src/agora/dozf.cpp (hunk 1 above)
Comment (rdoosty) [reply src/agora/dozf.cpp]: fixed!

File: src/agora/agora.cpp (hunk 2 above)
Comment (rdoosty) [reply src/agora/agora.cpp]: done!

**Pull request merged**